### PR TITLE
Added test_cpp_thread.py

### PR DESCRIPTION
**Pull request by jason-liu227**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

The tests that follow PR #2512.
PR #2512 NEEDS TO BE MERGED FIRST.

#### Which issue(s) this PR is related to:

Sub issues of 1249. linked to #2512

### CODE REVIEW

File: tests/profiler/test_cpp_thread.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        cpp.start_threads(self.ThreadCount, IterationCount, True)
+        self.check_trace(
+            {
+                "aten::add": [self.ThreadCount, "CPU"],
+            },
+            mem=True,
+        )
+
+
+instantiate_device_type_tests(CppThreadTest, globals(), only_for=("cpu"))
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py]: It should be a device level test like upstream tests: [instantiate_device_type_tests(CppThreadTest, globals(), only_for=("cuda", "xpu"))](https://github.com/pytorch/pytorch/blob/main/test/profiler/test_cpp_thread.py#L224)

File: tests/profiler/test_cpp_thread.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+else:
+    # cpp extensions use relative paths. Those paths are relative to
+    # this file, so we'll change the working directory temporarily
+    old_working_dir = os.getcwd()
+    os.chdir(os.path.dirname(os.path.abspath(__file__)))
+
+    cpp = torch.utils.cpp_extension.load(
+        name="profiler_test_cpp_thread_lib",
+        sources=[
+            "test_cpp_thread2.cpp",
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py]: Might be a typing mistake in the file name.

File: tests/profiler/test_cpp_thread.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            blueprint("stepping kineto profiler")
+            KinetoProfiler.step()
+
+    def emulateTraining(self, iteration: int, thread_id: int) -> None:
+        global device
+        # blueprint(f"training iteration {iteration} in thread {thread_id}")
+        torch_device = getattr(torch, device)
+        if not hasattr(torch_device, "synchronize"):
+            raise AssertionError(f"Device {device} does not have synchronize method")
+        sync_func = torch_device.synchronize
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py:79]: Since the device is only "cpu" it will be a "no op".

File: tests/profiler/test_cpp_thread.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            min_count = count * (ActivateIteration - 1)
+            dev = values[1]
+            filtered = filter(
+                lambda ev: ev.name == key
+                and str(ev.device_type) == f"DeviceType.{dev}",
+                event_list,
+            )
+
+            if mem:
+                mem_key = f"{self.device_type}_memory_usage"
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py:153]: This should check **device memory usage**, that's why the test should be written for spyre. It passes now because device is "cpu".

File: tests/profiler/test_cpp_thread.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                        "not enough event recorded",
+                    )
+
+    def test_with_enable_profiler_in_child_thread(self) -> None:
+        self.start_profiler(False)
+        cpp.start_threads(self.ThreadCount, IterationCount, True)
+        self.check_trace(
+            {
+                "aten::add": [self.ThreadCount, "CPU"],
+                "user_function": [self.ThreadCount, self.device_type.upper()],
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py]: This **user function**, which is context of record function, should get device side events abd be checked with `self.device_type.upper()` which should be equal to `DeviceType.PRIVATEUSE1`

File: tests/profiler/test_cpp_thread.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+            }
+        )
+
+    def test_without_enable_profiler_in_child_thread(self) -> None:
+        self.start_profiler(False)
+        cpp.start_threads(self.ThreadCount, IterationCount, False)
+        self.check_trace(
+            {
+                "aten::add": [1, "CPU"],
+                "user_function": [1, self.device_type.upper()],
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py]: This **user function**, which is context of record function, should get device side events abd be checked with `self.device_type.upper()` which should be equal to `DeviceType.PRIVATEUSE1`

File: tests/profiler/test_cpp_thread.py (hunk 1 above)
Comment (jason-liu227) [reply tests/profiler/test_cpp_thread.py]: Change to spyre?
instantiate_device_type_tests(CppThreadTest, globals(), only_for=("spyre"))

File: tests/profiler/test_cpp_thread.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        super().setUp()
+        global device
+        device = self.device_type
+
+        # Warmup pass
+        a = torch.ones(1, device=device)
+        b = torch.ones(1, device=device)
+        torch.add(a, b).cpu()
+        if hasattr(torch, device) and hasattr(getattr(torch, device), "synchronize"):
+            getattr(torch, device).synchronize()
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py]: Try adapting to upstream warmup style


```
    self.start_profiler(False)
    cpp.start_threads(1, IterationCount, False)
```

This runs a complete profiler lifecycle before each test:
1. Creates a full torch.profiler.profile object with the same schedule (wait=1, warmup=1, active=2)
2. Starts the C++ engine with 1 thread for 5 iterations
3. The profiler goes through: start → step → step → step → stop
4. on_trace_ready fires → self.set_trace() stores the trace on the class
5. All first-run initialization artifacts are now flushed

File: tests/profiler/test_cpp_thread.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        sync_func = torch_device.synchronize
+
+        with torch.autograd.profiler.record_function("user_function"):
+            a = torch.ones(1, device=device)
+            b = torch.ones(1, device=device)
+            torch.add(a, b).cpu()
+            sync_func()
+
+
+@unittest.skipIf(True, "Blocked on profiler infra - see #<issue>")
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py:88]: If `include torch/csrc/autograd/profiler_kineto.h`, works, then there's no need to skip.

File: tests/profiler/test_cpp_thread.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+            min_count = count * (ActivateIteration - 1)
+            dev = values[1]
+            filtered = filter(
+                lambda ev: ev.name == key
+                and str(ev.device_type) == f"DeviceType.{dev}",
+                event_list,
+            )
+
+            if mem:
+                mem_key = f"{self.device_type}_memory_usage"
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_cpp_thread.py:153]: f`"{self.device_type}_memory_usage"` will produce `"spyre_memory_usage"` or not needs to be verified that this is the actual field name PyTorch uses for PrivateUse1 memory attribution (it might be `"privateuse1_memory_usage"` instead, depending on how the profiler formats it). This is worth a quick verification once infra is ready.

**Comment by prasannabiswas-iitb** [comment]
What this test does (upstream intent): Just for the reference

The test validates that child C++ threads can dynamically join an active Kineto profiler session via enableProfilerInChildThread().
The flow is:
Python starts a profiler → C++ spawns N threads → each child thread checks isProfilerEnabledInMainThread() → if true, calls enableProfilerInChildThread() to join the session → after execution, Python asserts that events from ALL threads were captured (not just the main thread).

The host-side dispatch events (aten::add) always have DeviceType.CPU even when tensors are on a device, they represent the CPU thread's work of launching the op. The user_function annotation gets tagged with the device type of the active profiler session. The test_profile_memory variant validates cross-domain memory attribution (e.g., cuda_memory_usage appearing on a CPU dispatch event).

**Comment by jason-liu227** [comment]
With current test file, and the test_cpp_thread.cpp from PR #2512, of the 3 tests, only 1 pass. The other two tests fail due to `[W615 10:04:38.076977082 record_function.cpp:255] Requested callback is not found` and 
Screenshot From 2026-06-15 10-43-07

**Review by kaoutar55 — COMMENTED** [review]
The device-level test point, the cpu no-op, the only_for tuple, and the spyre_memory_usage field are already well covered in the existing threads, so I am not repeating those. One structural item I did not see raised:

File: tests/profiler/test_cpp_thread.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+if is_fbcode():
+    import caffe2.test.profiler_test_cpp_thread_lib as cpp  # @manual=//caffe2/test:profiler_test_cpp_thread_lib
+else:
+    # cpp extensions use relative paths. Those paths are relative to
+    # this file, so we'll change the working directory temporarily
+    old_working_dir = os.getcwd()
+    os.chdir(os.path.dirname(os.path.abspath(__file__)))
+
+    cpp = torch.utils.cpp_extension.load(
```
Comment (kaoutar55) [inline tests/profiler/test_cpp_thread.py:33]: `cpp_extension.load(sources=[...])` runs at import time here, but `test_cpp_thread.cpp` is added in #2512, which is not merged, so that source is not in the tree. The class-level `skipIf` does not prevent this: the `load()` call runs during import and fails to compile, which breaks test collection for the whole file. So this PR can only pass CI after #2512 merges. It should be marked blocked on #2512, and ideally the `load()` should be wrapped so a missing source skips the test instead of erroring during collection.

Comment (jason-liu227) [reply tests/profiler/test_cpp_thread.py:33]: Yes that is true, it was suggested to have 2 PRs for the two items.
However when testing on a Pod, both files exist on the same branch and will skip the test.

So yes #2512 will need to be merged before this PR.

**Comment by ppnaik1890** [comment]
Hi @jason-liu227 
#2512 was closed. Since this had a dependency on it can you elaborate on how we remove it? 
@prasannabiswas-iitb can you please re-review this PR.

**Comment by jason-liu227** [comment]
@ppnaik1890 
The blocker for this issue is a lack of support that me and @prasannabiswas-iitb agreed on. 
There is a cpu- sided annotation that makes it so its unable to direct to spyre even though spyre is availabe.
